# Problème de répartition équitable de bonus

## Contexte

Dans le cadre de l'évaluation des performances de ses élèves, un enseignant souhaite ajuster les moyennes afin de garantir que tous les élèves valident la matière enseignée. Le tableau ci-dessous présente les moyennes actuelles des élèves, notées sur 20 :

|N°|Noms|Moyennes|
|:-:|:-:|:-:|
|1|Aya|8|
|2|Souley|15|
|3|Juste|19|
|4|Kouakou|4|
|5|David|9|
|6|Ange|13|

Dans cet établissement, la moyenne minimale requise pour valider une matière est fixée à 8 sur 20. Or, Kouakou, ayant obtenu une moyenne de 4, n’atteint pas ce seuil. Soucieux de la réussite de tous ses élèves, l'enseignant envisage d'ajouter un bonus de +4 à la moyenne de Kouakou pour le ramener à 8, le seuil de validation.

Cependant, cette décision suscite des mécontentements parmi les autres élèves, qui estiment mériter un ajustement similaire pour maintenir l'équité des résultats. Ces élèves souhaitent ainsi recevoir un bonus proportionnel tout en conservant les écarts existants entre leurs moyennes. Néanmoins, l'application d'un bonus de +4 à la moyenne de Juste, qui est actuellement de 19, entraînerait une moyenne de 23, ce qui excède le maximum autorisé de 20, rendant la situation incohérente.

Le défi consiste donc à redistribuer les bonus de manière équitable, en permettant à chaque élève de valider la matière tout en respectant les contraintes suivantes :
- La moyenne de validation doit être comprise entre 8 et 20.
- Les moyennes actuelles des élèves varient entre 4 et 19.
- La plage de bonus applicables est donc comprise entre +1 et +4.
- La difficulté : faire en sorte que [4, 19] + [1, 4] -> [8, 20]

Cette situation illustre un problème plus général de répartition équitable de ressources dans un groupe tout en respectant des limites prédéfinies.

## Statistiques de départ

Importation des modules

In [2]:
import numpy as np

Déclaration des variables

In [3]:
liste_notes = [8, 15, 19, 4, 9, 13]
notes = np.array(np.sort(liste_notes), dtype=float)

VALIDATION_MIN = 8 # note minimale de validation
VALIDATION_MAX = 20 # note maximale de validation
note_min = int(notes.min()) # note minimale de la classe
note_max = int(notes.max()) # note maximale de la classe
bonus_max = VALIDATION_MIN - note_min # bonus maximal possible à ajouter
bonus_min = min(bonus_max, VALIDATION_MAX - note_max) # bonus maximal possible à ajouter

Quelques statistiques sur les notes des élèves

In [4]:
notes.mean() # la moyenne de la classe

11.333333333333334

In [5]:
notes.var() # la variance de la classe

24.222222222222218

In [6]:
notes.std() # l'écart-type de la classe

4.9216076867444665

## Tentatives de solution

### Solution N°1

In [7]:
liste_bonus_1 = np.linspace(bonus_max, bonus_min, notes.size) # liste des bonus à appliquer
liste_bonus_1

array([4. , 3.4, 2.8, 2.2, 1.6, 1. ])

In [10]:
resultat = [note + bonus for note, bonus in zip(notes, liste_bonus_1) if note in notes]
notes_1 = np.array(resultat)
notes_1

array([ 8. , 11.4, 11.8, 15.2, 16.6, 20. ])

Après application des bonus, nous obtenons ainsi le tableau suivant :

|N°|Noms|Notes initiales|Bonus|Notes finales|
|:-:|:-:|:-:|:-:|:-:|
|1|Juste|19|+1|20|
|2|Souley|15|+1.6|16.6|
|3|Ange|13|+2.2|15.2|
|4|David|9|+2.8|11.8|
|5|Aya|8|+3.4|11.4|
|6|Kouakou|4|+4|8|
||**Moyennes**|**11.33**|**+2.5**|**13.83**|

Quelques statistiques de la solution 1

In [11]:
notes_1.mean() # la moyenne de la classe

13.833333333333334

In [12]:
notes_1.var() # la variance de la classe

15.272222222222224

In [13]:
notes_1.std() # l'écart-type de la classe

3.9079690661803124

### Solution N°2

In [19]:
liste_temp_2 = np.linspace(note_min, note_max, note_max - note_min + 1)
liste_temp_2

array([ 4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13., 14., 15., 16.,
       17., 18., 19.])

In [20]:
liste_bonus_2 = np.linspace(bonus_max, bonus_min, note_max - note_min + 1)
liste_bonus_2

array([4. , 3.8, 3.6, 3.4, 3.2, 3. , 2.8, 2.6, 2.4, 2.2, 2. , 1.8, 1.6,
       1.4, 1.2, 1. ])

In [21]:
resultat = [note + bonus for note, bonus in zip(liste_temp_2, liste_bonus_2) if note in notes]
notes_2 = np.array(resultat)
notes_2

array([ 8. , 11.2, 12. , 15.2, 16.8, 20. ])

Après application des bonus, nous obtenons ainsi le tableau suivant :

|N°|Noms|Moyennes initiales|Bonus|Moyennes finales|
|:-:|:-:|:-:|:-:|:-:|
|1|Juste|19|+1|20|
|2|Souley|15|+1.8|16.8|
|3|Ange|13|+2.2|15.2|
|4|David|9|+3|12|
|5|Aya|8|+3.2|11.2|
|6|Kouakou|4|+4|8|
||**Moyennes**|**11.33**|**+2.53**|**13.87**|

Statistiques de la solution N°2

In [22]:
notes_2.mean() # la moyenne de la classe

13.866666666666667

In [23]:
notes_2.var() # la variance de la classe

15.502222222222224

In [24]:
notes_2.std() # l'écart-type de la classe

3.937286149395574

## Interprétation des résultats

|Méthodes|Naturelle|Solution 1|Solution 2|
|:-:|:-:|:-:|:-:|
|Note minimale|4|8|8|
|Note maximale|19|20|20|
|Moyenne|11,33|13,83|13,86|
|Variance|24.22|15.27|15.5|
|Ecart-type|4.92|3.91|3.94|

1. La seconde méthode améliore mieux la moyenne des élèves que la première.
2. La première méthode réduit les variances dans l'ensemble. Ce qui est plus avantagieux pour les derniers mais désavantagieux pour les premiers.